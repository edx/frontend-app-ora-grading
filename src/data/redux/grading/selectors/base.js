import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import * as module from './base';

export const rootSelector = ({ grading }) => grading;
export const mkSimpleSelector = (key) => createSelector(
  [module.rootSelector],
  (grading) => grading[key],
);
export const rootKeys = StrictDict({
  selected: 'selected',
  activeIndex: 'activeIndex',
  current: 'current',
  gradeData: 'gradeData',
  gradingData: 'gradingData',
});

export const simpleSelectors = {
  selected: mkSimpleSelector(rootKeys.selected),
  activeIndex: mkSimpleSelector(rootKeys.activeIndex),
  current: mkSimpleSelector(rootKeys.current),
  gradeData: mkSimpleSelector(rootKeys.gradeData),
  gradingData: mkSimpleSelector(rootKeys.gradingData),
};

/**
 * returns the length of the list of selected submissions
 * @return {number} selected submission list length
 */
export const selectionLength = createSelector(
  [module.simpleSelectors.selected],
  (selected) => selected.length,
);

export default {
  ...simpleSelectors,
  selectionLength,
};
