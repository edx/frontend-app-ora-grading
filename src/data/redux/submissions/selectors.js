import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { lockStatuses } from 'data/services/lms/constants';

export const simpleSelectors = {
  allSubmissions: state => state.submissions.allSubmissions,
};

/**
 * Returns the submission list in default order for the table.
 */
export const listData = createSelector(
  [simpleSelectors.allSubmissions],
  (allSubmissions) => {
    const submissionIds = Object.keys(allSubmissions);
    const submissionList = submissionIds.map(id => {
      const { gradeStatus, lockStatus, ...rest } = allSubmissions[id];
      const gradingStatus = (lockStatus === lockStatuses.unlocked ? gradeStatus : lockStatus);
      return { gradingStatus, ...rest };
    });
    return _.sortBy(
      submissionList,
      ['submissionDate'],
    );
  },
);

export default StrictDict({
  ...simpleSelectors,
  listData,
});
