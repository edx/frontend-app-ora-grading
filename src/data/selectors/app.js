import { StrictDict } from 'utils';

export const simpleSelectors = {
  showReview: state => state.app.showReview,
  showRubric: state => state.app.showRubric,
  grading: state => state.app.grading,
  courseMetadata: state => state.app.courseMetadata,
  oraName: state => state.app.oraMetadata.name,
  oraPrompt: state => state.app.oraMetadata.prompt,
  oraTypes: state => state.app.oraMetadata.type,
};

export default StrictDict({
  ...simpleSelectors,
});
