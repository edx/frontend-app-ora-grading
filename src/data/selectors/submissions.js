import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';

const simpleSelectors = {
  list: state => state.submissions.list,
  selected: state => state.submissions.selected,
  activeIndex: state => state.submissions.activeIndex,
  selectedResponse: state => state.submissions.current.response,
};

export const selectionIndex = createSelector(
  [simpleSelectors.list],
  (list, { learnerId }) => list.selected.indexOf(learnerId),
);

export const listData = createSelector(
  [simpleSelectors.list],
  (list) => _.sortBy(
    Object.values(list),
    ['learnerId'],
  ),
);

export const selectionLearnerId = (state, { index }) => (
  simpleSelectors.selected(state)[index].learnerId
);

export const currentSelection = createSelector(
  [simpleSelectors.selected, simpleSelectors.activeIndex],
  (list, activeIndex) => list[activeIndex],
);

export const selectedLearnerId = createSelector(
  [currentSelection],
  (selection) => selection.learnerId,
);

export const prevLearnerId = createSelector(
  [simpleSelectors.selected, simpleSelectors.activeIndex],
  (list, activeIndex) => {
    if (activeIndex > 0) {
      return list[activeIndex - 1].learnerId;
    }
    return null;
  },
);
export const nextLearnerId = createSelector(
  [simpleSelectors.selected, simpleSelectors.activeIndex],
  (list, activeIndex) => {
    if (activeIndex < list.length - 1) {
      return list[activeIndex + 1].learnerId;
    }
    return null;
  },
);

export default StrictDict({
  ...simpleSelectors,
  currentSelection,
  selectionIndex,
  listData,
  selectedLearnerId,
  nextLearnerId,
  prevLearnerId,
});
