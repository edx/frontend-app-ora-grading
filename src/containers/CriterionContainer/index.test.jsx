import React from 'react';
import { shallow } from 'enzyme';

import selectors from 'data/selectors';
import { CriterionContainer, mapStateToProps } from '.';

jest.mock('components/InfoPopover', () => 'InfoPopover');
jest.mock('./RadioCriterion', () => 'RadioCriterion');
jest.mock('./CriterionFeedback', () => 'CriterionFeedback');
jest.mock('./ReviewCriterion', () => 'ReviewCriterion');

jest.mock('@edx/paragon', () => ({
  Form: {
    Group: () => 'Form.Group',
    Label: () => 'Form.Label',
  },
}));

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      rubric: {
        criterionConfig: jest.fn((...args) => ({ _criterionConfig: args })),
      },
    },
    grading: {
      selected: {
        gradeStatus: jest.fn((...args) => ({ selectedGradeStatus: args })),
      },
    },
  },
}));

describe('Criterion Container', () => {
  const props = {
    isGrading: true,
    orderNum: 1,
    config: {
      prompt: 'prompt',
      name: 'random name',
      feedback: 'feedback mock',
      options: [
        {
          explanation: 'explaination',
          feedback: 'option feedback',
          label: 'this label',
          name: 'option name',
          points: 2,
        },
        {
          explanation: 'explaination 2',
          feedback: 'option feedback 2',
          label: 'this label 2',
          name: 'option name 2',
          points: 1,
        },
      ],
    },
    gradeStatus: 'ungraded',
  };
  let el;
  beforeEach(() => {
    el = shallow(<CriterionContainer {...props} />);
  });

  describe('snapshot', () => {
    test('is ungraded and is grading', () => {
      expect(el).toMatchSnapshot();
    });

    test('is ungraded and is not grading', () => {
      el.setProps({
        isGrading: false,
      });
      expect(el).toMatchSnapshot();
    });

    test('is graded and is not grading', () => {
      el.setProps({
        isGrading: false,
        gradeStatus: 'graded',
      });
      expect(el).toMatchSnapshot();
    });
  });

  describe('component', () => {
    test('rendering', () => {
      expect(el.isEmptyRender()).toEqual(false);
      const optionsEl = el.find('.help-popover-option');
      expect(optionsEl.length).toEqual(props.config.options.length);
      optionsEl.forEach((optionEl, i) => {
        expect(optionEl.key()).toEqual(props.config.options[i].name);
        expect(optionEl.text()).toContain(props.config.options[i].explanation);
      });
    });

    test('is ungraded and is grading', () => {
      const rubricCritera = el.find('.rubric-criteria');
      expect(rubricCritera.children(0).name()).toEqual('RadioCriterion');
    });

    test('is ungraded and is not grading', () => {
      el.setProps({
        isGrading: false,
      });
      const rubricCritera = el.find('.rubric-criteria');
      expect(rubricCritera.children(0).name()).toEqual('ReviewCriterion');
    });

    test('is graded and is not grading', () => {
      el.setProps({
        isGrading: false,
        gradeStatus: 'graded',
      });
      const rubricCritera = el.find('.rubric-criteria');
      expect(rubricCritera.children(0).name()).toEqual('RadioCriterion');
    });
  });

  describe('mapStateToProps', () => {
    const testState = { abitaryState: 'some data' };
    const additionalArgs = { orderNum: props.orderNum };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState, additionalArgs);
    });
    test('selectors.app.rubric.criterionConfig', () => {
      expect(mapped.config).toEqual(
        selectors.app.rubric.criterionConfig(testState, additionalArgs),
      );
    });

    test('selectors.grading.selected.gradeStatus', () => {
      expect(mapped.gradeStatus).toEqual(
        selectors.grading.selected.gradeStatus(testState),
      );
    });
  });
});
