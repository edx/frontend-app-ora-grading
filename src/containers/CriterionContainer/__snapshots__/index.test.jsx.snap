// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Criterion Container snapshot 1`] = `
<Group>
  <Label
    className="criteria-label"
  >
    <span
      className="criteria-title"
    >
      prompt
    </span>
    <InfoPopover>
      <div
        className="help-popover-option"
        key="option name"
      >
        <strong>
          this label
        </strong>
        <br />
        explaination
      </div>
      <div
        className="help-popover-option"
        key="option name 2"
      >
        <strong>
          this label 2
        </strong>
        <br />
        explaination 2
      </div>
    </InfoPopover>
  </Label>
  <div
    className="rubric-criteria"
  >
    <RadioCriterion
      isGrading={true}
      orderNum={1}
    />
  </div>
  <CriterionFeedback
    isGrading={true}
    orderNum={1}
  />
</Group>
`;